code/modules/dwarven/dwarven_structures.dm

```diff
@@ -0,0 +1,428 @@
+		/datum/material/diamond,
+		/datum/material/uranium,
+		/datum/material/plasma,
+		/datum/material/bluespace,
+		/datum/material/bananium,
+		/datum/material/titanium,
+		/datum/material/runite,
+		/datum/material/plastic,
+		/datum/material/adamantine,
+		/datum/material/mythril,
+		/datum/material/wood,
+		/datum/material/dwarven,
+		)
+	//temporary solution til alloys become its own datum.
+	var/static/list/banned_alloys = list(
+		/obj/item/stack/sheet/mineral/plastitanium,
+		/obj/item/stack/sheet/plasteel,
+		/obj/item/stack/sheet/plasmaglass,
+		/obj/item/stack/sheet/rglass,
+		/obj/item/stack/sheet/plasmarglass,
+		/obj/item/stack/sheet/titaniumglass,
+		/obj/item/stack/sheet/plastitaniumglass
+		)
+	var/datum/material/current_mat
+
+/obj/structure/destructible/dwarven/anvil/Initialize()
+	. = ..()
+	//temporary solution til alloys become its own datum.
+	var/list/allowed_materials = list()
+	for(var/i in typecacheof(/obj/item/stack))
+		if(!(i in banned_alloys))
+			allowed_materials += i
+
+	AddComponent(/datum/component/material_container,allowed_types, 200000, TRUE, allowed_materials ,null, _after_insert=CALLBACK(src, .proc/AfterMaterialInsert))
+
+/obj/structure/destructible/dwarven/anvil/AltClick(mob/user)
+	if(!istype(user) || !user.canUseTopic(src, BE_CLOSE))
+		return
+	if(state == 5)
+		state = 0
+	else
+		state++
+
+/obj/structure/destructible/dwarven/anvil/proc/AfterMaterialInsert(_item_inserted, id_inserted, amount_inserted)
+	var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+	for(var/datum/material/M in materials.materials)
+		if(M != id_inserted)
+			materials.retrieve_sheets(materials.materials[M],M,drop_location())
+
+	current_mat = id_inserted
+
+/obj/structure/destructible/dwarven/anvil/attack_hand(mob/user)
+	var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+	materials.retrieve_all()
+
+	. = ..()
+
+
+/obj/structure/destructible/dwarven/anvil/attacked_by(obj/item/I, mob/living/user)
+	var/mob/living/carbon/human/H = user
+	if(!H.has_language(/datum/language/dwarven))
+		to_chat(H, "<span class='notice'>You don't understand how to operate this ancient machinery!</span>")
+		return
+	if(istype(I,/obj/item/dwarven/mallet) && current_mat != null)
+		handle_mallet(user,current_mat)
+		return
+	. = ..()
+///This proc contains everything that happens when you hit the parent with mallet hence the name handle_mallet
+/obj/structure/destructible/dwarven/anvil/proc/handle_mallet(mob/living/user,datum/material/S)
+	var/list/materials_used = list()
+
+	var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+
+	if(!do_after(user, 60, target = src))
+		return
+
+	var/obj/item/item_built = crafting_list[state+1]
+	if(!materials.has_enough_of_material(S,crafting_list[item_built]))
+		to_chat(user, "<span class='alert'>Not enough materials for this operation.</span>")
```

Comment:
yeah after some nitwit uses the anvil with his hand, removing all the ores mid operation